### ISSUE TITLE
[TabArena] Benchmarking PerpetualBoosting; Verifying our Implemenetation

### ISSUE THREAD

**Issue by LennartPurucker** [perpetual-ml/perpetual#66]
Heyho,

We are working on benchmarking your interesting model on [TabArena](https://tabarena.ai). I have implemented and benchmarked a first version of your model in this pull request: https://github.com/autogluon/tabrepo/pull/170

The PR still has a few open TODOs and requires your input to verify its correctness. 

In particular, I had several issues with time and memory limits not being strictly enforced, and the lack of support for passing our own validation data. Thus, I expect the current results to be suboptimal for PerpetualBoosting. I detailed some of my problems in the code itself. 

I'm happy to discuss this further here or in the PR about how to maximize the performance of your model. 
Feel free to test the implementation yourself or propose changes I should incorporate. 

Once everyone is happy with the code, I will run the config (with HPO) on all of TabArena and add its results to the public leaderboard. 

Thank you!

**Comment by LennartPurucker** [comment]
Thank you for the details! 

> https://github.com/deadsoul44/automlbenchmark/blob/master/frameworks/PerpetualBooster/exec.py?rgh-link-date=2025-06-26T14%3A46%3A54.000Z#L45

From this I understand that you want to use `budget=0.1` in benchmarks. 


> memory_limit for that dataset.

The dataset was this one: https://www.openml.org/search?type=task&sort=runs&id=363711
When I set the limit to 32 GB, I had to give 64 GB to not run OOM. When I set the limit to 64 GB, the process used 90ish GB of memory. In any case, it runs into an OOM or hangs indefinitely during the prediction later.  

Note that I am benchmarking on SLURM, which, depending on the node configuration, has shared resources that are controlled through affinity and cgroups. I am unsure how RUST handles this edge case. For Python, this usually requires annoying workarounds and special logic. 

> timeout is a hard limit to stop boosting. It is not the same thing as time_budget in other automl frameworks. [...] timeout is checked after completing each boosting round. The algorithm may slightly exceed the timeout

Is there any way to control the timeout and treat it like a callback in other models (or AutoML) frameworks? 
We require that models strictly obey a time limit. 

Moreover, other models usually implement the time limit such that they do not start another iteration if the can estimate that the iteration would take longer than the amount of time is left. Would it be possible to support similar logic in your code? 

> I suggest to run the algorithm for different budget levels, e.g. 0.1, 0.2, 0.5, 1.0, 1.5, 2.0. 

Would you consider this a hyperparameter that we should search over? Or how should I know for a new dataset and any the user-defined time limit how to set the budget? 

I would expect the following behavior: set it to the highest possible budget and then run until a time limit is reached. 

> It is not needed to set stopping_rounds. It has almost no effect on results.

We can also set it to `inf` if needed (or a very high int).

> No need to set memory_limit per thread. It is always total memory allocated.

In my experiments with the code, the memory limit was not controlled across threads.

If I set num_threads>1, then the whole training process was killed for some datasets because the process used more than 32 GB of RAM across all threads.

Note that this is likely related to SLURM / or any other shared resources setting, where certain Linux functions might tell you that you have more RAM on the node but the cgroup of the process will kill the process once it goes over the actual limit. 

> # at this stage, num_cpus == 1 anyhow. This shouldn't be the case. 

No worries, your parallelism works fine here. We fit each fold on one CPU in the case of parallel fold fitting, so for the current resource settings, I know that num_cpus would be 1. 

> PerpetualBooster does not use validation data

Alright. I would like to note that this very likely will put you at a disadvantage compared to other models.
Similar to how we treat foundation models, we could also set it such that we refit the model on all data after cross-validation. 

If you think it would be more optimal to refit on all data than using cross-validation ensembles. 

> No GPU support yet.

That is fine. Models do not needto support GPU. 

> Seed

Would there be an option to specify the seed somehow? We would like to guarnatee deterministic behavior

**Comment by deadsoul44** [comment]
> From this I understand that you want to use `budget=0.1` in benchmarks.

No, I suggest running with different `budget` levels (0.1, 0.2, 0.5, 1.0, 1.5, 2.0) and take the best result.

> The dataset was this one: https://www.openml.org/search?type=task&sort=runs&id=363711 When I set the limit to 32 GB, I had to give 64 GB to not run OOM. When I set the limit to 64 GB, the process used 90ish GB of memory. In any case, it runs into an OOM or hangs indefinitely during the prediction later.

This problem is probably due to memory allocation with categorical data. We should fix this in a future release.
I suggest decreasing memory limit. For example, if 32 GB is what you want, set the limit to 16 GB for that dataset. 

> Note that I am benchmarking on SLURM, which, depending on the node configuration, has shared resources that are controlled through affinity and cgroups. I am unsure how RUST handles this edge case. For Python, this usually requires annoying workarounds and special logic.

I don't have any experience using SLURM but we use Rayon for parallelism. Rust's Rayon library is generally well-behaved in SLURM environments.

> Moreover, other models usually implement the time limit such that they do not start another iteration if the can estimate that the iteration would take longer than the amount of time is left. Would it be possible to support similar logic in your code?

This is possible but it will not be robust in any way. Moreover, this is not needed in this benchmark. Because, you will run the algorithm with different budget levels and take the best result. Best result case will strictly obey the time limit. 

> > I suggest to run the algorithm for different budget levels, e.g. 0.1, 0.2, 0.5, 1.0, 1.5, 2.0.
> 
> Would you consider this a hyperparameter that we should search over? Or how should I know for a new dataset and any the user-defined time limit how to set the budget?

No, budget is not a hyperparameter. Higher budget means better results. `timeout` is only a safety measure in case of a noisy data to prevent running algorithm forever. You are thinking in terms of classical automl frameworks. Automl frameworks randomly throw algorithms at data for predefined period of time. 

> I would expect the following behavior: set it to the highest possible budget and then run until a time limit is reached.

I suggest running the algorithm with different budget levels (0.1, 0.2, 0.5, 1.0, 1.5, 2.0) and take the best result. The best result will be the case before the one that hits the time limit. Use the same time limit that you use for other algorithms / frameworks. 

> > It is not needed to set stopping_rounds. It has almost no effect on results.
> 
> We can also set it to `inf` if needed (or a very high int).

Simply ignore it. Do not set anything. The algorithm has a default value.

> > No need to set memory_limit per thread. It is always total memory allocated.
> 
> In my experiments with the code, the memory limit was not controlled across threads.
> 
> If I set num_threads>1, then the whole training process was killed for some datasets because the process used more than 32 GB of RAM across all threads.
> 
> Note that this is likely related to SLURM / or any other shared resources setting, where certain Linux functions might tell you that you have more RAM on the node but the cgroup of the process will kill the process once it goes over the actual limit.

This can be due to data or SLURM. In these cases, try with a lower memory limit.

> > PerpetualBooster does not use validation data
> 
> Alright. I would like to note that this very likely will put you at a disadvantage compared to other models. Similar to how we treat foundation models, we could also set it such that we refit the model on all data after cross-validation.
> 
> If you think it would be more optimal to refit on all data than using cross-validation ensembles.

This should be an advantage. PerpetualBooster does not need cross-validation. Fit with all the data. Test the model on test data. In other words, a single fit run is enough for PerpetualBooster. 

> > Seed
> 
> Would there be an option to specify the seed somehow? We would like to guarantee deterministic behavior and ideally be able to adapt the seed if needed.

Current version is deterministic. It has a default seed value. 

Let me know if you have any other comments or questions.

**Comment by deadsoul44** [comment]
There is no need for cross-validation since PerpetualBooster has a built-in mechanism to prevent overfitting. Use all the training data, and fit a single model for each budget level. 

I suggest running with different budget levels (0.1, 0.2, 0.5, 1.0, 1.5, 2.0) and take the model before the one that hits the time limit. For example, if `budget=1.0` hits the time limit, take the model with `budget=0.5`.  

Use the model above for testing.

**Comment by LennartPurucker** [comment]
Okay, I think we are still misaligned. 

Let me explain where this approach fails to work in a real-world pipeline, and please correct me where I am misunderstanding it:

1. I can create the final model that I want to deploy by fitting one model. However, I must estimate beforehand how good this model is. For example, if I cross-validate PerpetualBooster and CatBoost, and CatBoost achieves a higher validation score, I would choose CatBoost for deployment instead of PerpetualBooster (or vice versa). Without cross-validation, I cannot even make this **necessary** comparison before deployment. Likewise, how would I know that PerpetualBooster meets the minimal required confidence, calibration, or accuracy to fulfill my application constraints? 

2. I have a time limit of 1 hour. If `budget=1.0` hits the time limit and stops, I cannot train another model, as my allotted time to train a model has expired. I cannot restart the training as I do not know a priori if some budget will hit the time limit. What is possible, albeit a somewhat unconventional alternative, is to have a loop in the train function that first fits with a budget of 0.1, then with a budget of 0.2, and so on. However, this seems inefficient, as we are wasting fit time, and later budgets might receive less total time. It would seem logical to me that if I call the model with budget 2.0 and give it a time limit, it should be, to some extent, the superset of the model of budget 0.1 to 2.0.  If you want to have such a loop, I am happy to implement it.

**Comment by deadsoul44** [comment]
I think we are converging to a common point :)

1. Check our examples to see how we benchmark PerpetualBooster against LightGBM:
https://github.com/perpetual-ml/perpetual/blob/main/python-package/examples/benchmark_lgbm.py
https://github.com/perpetual-ml/perpetual/blob/main/python-package/examples/benchmark_perpetual.py
Cross-validation is usually applied to find the best hyperparameters of an algorithm. Check scikit-learn docs as an example (especially the first image):
https://scikit-learn.org/stable/modules/cross_validation.html
They split the data into train and test sets and use the train data for cross validation. They use the test set for the final evaluation. 
We use a similar approach in our benchmarks but we split the data five times with different seeds to prevent splitting spoil the results.
Since we don't have to worry about finding the best hyperparameters for PerpetualBooster, we use all the train set for fitting the model.
2. Increasing budget increases fit time for PerpetualBooster. This is similar to number of boosting rounds for other GBMs. Your concern about not knowing fit time is also valid for other GBMs. How do you know the required number of boosting rounds to fit in 1 hour time limit?

**Comment by deadsoul44** [comment]
1. As also suggested per scikit-learn docs; final evaluation including confidence, calibration, or accuracy should be checked with test data. 

> From prior work and TabArena data, we observed that for most models, a cross-validation ensemble outperforms refitting and is more time-efficient.

Fitting on all data is more efficient for PerpetualBooster.

2. In this case, you are effectively searching for a maximum number of boosting rounds which fill fit in the predefined time limit. When you are benchmarking another GBM with hyperparameter optimization with 100 trials, how do you define time limit per fitting? 1 h / 100? If this is the case, some trials are probably finishing before time limit which will limit full utilization of the available time. 

Having to define a time limit is the downside of these benchmarks. For a user, it is not possible to estimate time limit beforehand. PerpetualBooster enables users to define a unitless budget. The algorithm stops itself without validation data. It has a built-in validation mechanism.

**Comment by LennartPurucker** [comment]
> test data.

You do not have the labeled test data in real-world use cases to make such statements. 
We only have validation data.

> how do you define time limit per fitting? 

We use 1 hour per fit. And run 200 configurations (up to 200 hours) for each HP.
This is largely a result of academic practice, but it provides a good estimate of the model's performance.

Most configurations stop way before the 1-hour time limit in our benchmark, given our dataset sizes.
At most 1% of runs exceed the time limit (see the Appendix of our paper). For these, we can increase the time limit to 2 hours and be fine.

> Having to define a time limit is the downside of these benchmarks.

I strongly disagree. A user cannot simply run a job and expect it to end randomly at some point. Any real-world usage is always limited in resources, usually expressed in the time one is willing to wait (or spend on compute).

---

We do not seem to have converged on a common understanding of machine learning so far. Moreover, given that we are running in circles, I suggest the following: I re-run PerpetualBooster with the points we were able to agree on, and stay with the rest as is. 

This will likely put it at a disadvantage compared to other models, as I mentioned earlier. Other models support the requirements I put forward above because they converged to the same understanding of machine learning that we aim to evaluate with TabArena.

**Comment by dholzmueller** [comment]
Hi, another TabArena co-author here. I'll try to clarify the discussion around validation/test data:
Besides training data (T), you may need non-training data subsets for 
(A) early stopping, 
(B) hyperparameter optimization, 
(C) algorithm selection (finding out what is the best algorithm to apply to your problem), and 
(D) getting an unbiased estimate of the performance of the final model. 

@deadsoul44 you are saying that A=B=validation, C=test. In contrast, in AutoML, one would typically use A=B=C=validation, so one doesn't have to hold out separate sets for this, while one would say D=test.
Regardless of whether you need D or not (in the benchmark we do), in practice you will have a fixed set of data you can allocate to T+A+B+C. If you want C to be a separate set, you will have less data for T+A+B. Since we want to simulate a scenario of a user that wants to do algorithm selection, we need predictions of PerpetualBoosting on C, so it will not get C for training, since train scores would be too unreliable for algorithm selection. The option that Lennart described is to fit PerpetualBoosting in cross-validation on T+A+B to get scores on C, then fit it again on T+A+B+C so it has more data. Of course this takes ~8x more time than fitting it directly on T+A+B+C, but this is the only way we can get validation scores on C.

**Comment by deadsoul44** [comment]
> You do not have the labeled test data in real-world use cases to make such statements.
We only have validation data.

In real-world, we create separate sets from data that we have. Scikit-learn reference shows one way to do it and they call one set as <ins>test data</ins>.
https://scikit-learn.org/stable/modules/cross_validation.html

> Most configurations stop way before the 1-hour time limit in our benchmark

There is something wrong here. If they stop before hitting time limit, it means that it is hitting lower bound of learning rate if you are optimizing it for GBMs since you are setting number of boosting rounds to inf or very large number. 

Another problem is that if most configurations stop before time limit, your time limit is ineffective to compare algorithms. 

> given our dataset sizes.

You are probably not using large datasets. We are using the largest datasets from OpenML on our benchmarks against AutoGluon. I suggest using large datasets to represent real world. 

> A user cannot simply run a job and expect it to end randomly at some point. Any real-world usage is always limited in resources, usually expressed in the time one is willing to wait (or spend on compute).

A user cannot simply define a time limit for a dataset and hope to get the best result. A dataset may need more time than specified time limit or a pre-defined time limit can be overkill for a dataset. You can specify number of boosting rounds to manage time to train a GBM. Moreover, your time limit on your benchmarks is ineffective. 

> We do not seem to have converged on a common understanding of machine learning so far. 

Because, you are ignoring machine learning fundamentals to enforce your approach. 

> Moreover, given that we are running in circles

Because, I am referencing scikit-learn docs, but somehow you are ignoring it. 

> I suggest the following: I re-run PerpetualBooster with the points we were able to agree on, and stay with the rest as is.

I suggest revising your benchmark method before including PerpetualBooster.

> This will likely put it at a disadvantage compared to other models, as I mentioned earlier.

Being at an advantage or disadvantage on a susceptible benchmark does not mean anything. 

> Other models support the requirements I put forward above because they converged to the same understanding of machine learning that we aim to evaluate with TabArena.

Obviously me and thousands of scikit-learn contributors do not have the same understanding with you and your others.

**Comment by dholzmueller** [comment]
Regarding the time limit: Our time limit is there so that methods don't take an eternity to run. Think of a user that prefers fast methods but wants to wait at most an hour (for a single configuration). We report the runtimes of algorithms, so the fact that they mostly don't take an hour is a good thing that is honored by the benchmark, because it means that these methods are reasonably fast to run. Our goal is not to answer the question "which algorithm can achieve the best performance within an hour?". For boosted trees, this means that we take an n_estimators that is sufficiently high such that they can come close to their optimal performance, but don't run forever. We could take a different n_estimators and different learning_rate space and get a different tradeoff, of course. As discussed, we have the following options for PerpetualBoosting:
- take the different budgets as different hyperparameter configs (each of them gets a time limit of 1hr). One could just take the largest budget that doesn't hit the time limit in the end as you suggested, but since we simulate the scenario where one needs validation scores anyway (see previous post), we also have validation scores which means we could just take the budget with the best validation score (this would also be easier to do in the code, since it is the same as for other methods). The problem with this is that we might run multiple budgets that hit the time limit, which gives a time estimate that is larger than if we would run the budgets in increasing order until we hit the time limit.
- do a for-loop increasing the budget until we hit the time limit, then take the largest budget that didn't hit the time limit. Of course this means that the total time limit for all budgets will be an hour, which is more restrictive than the other option. I think at this point we would basically be benchmarking an AutoML method that tries to exhaust a time limit and compare it to regular ML methods that don't try to exhaust a time limit but of which we can run multiple configurations that together can go far beyond 1h. This comparison is a bit apples to oranges but it might still be interesting. (Question: can we expect the runtime to be roughly proportional to the budget? In this case we could avoid running the last budget that would likely hit the time limit. Otherwise I guess the problem remains that PerpetualBoosting doesn't enforce the time limit as strictly as other methods.)

In general, either solution could be acceptable but won't be great IMHO.
